build-logic/publication/src/main/kotlin/convention.sonatype.gradle.kts

```diff
@@ -1,32 +1,74 @@
 @file:Suppress("UnstableApiUsage")
 
+import com.avito.android.publish.CreateStagingRepositoryTask
+
 plugins {
     id("convention.publish-base")
     signing
 }
 
+val ossrhUsername: Provider<String> = providers.gradleProperty("avito.ossrh.user")
+    .forUseAtConfigurationTime()
+
+val ossrhPassword: Provider<String> = providers.gradleProperty("avito.ossrh.password")
+    .forUseAtConfigurationTime()
+
+val ossrhStagingProfileId: Provider<String> = providers.gradleProperty("avito.ossrh.stagingProfileId")
+    .forUseAtConfigurationTime()
+
+val sonatypeRepoName = "SonatypeReleases"
+
+val repositoryUrlOutputFilePath: Provider<RegularFile> = rootProject.layout.buildDirectory.file("sonatype-repo.id")
+
+val buildId: Provider<String> = providers.gradleProperty("teamcityBuildId").forUseAtConfigurationTime()
+
+val createStagingRepositoryTask: TaskProvider<CreateStagingRepositoryTask> = with(rootProject.tasks) {
+    val createStagingTaskName = "createSonatypeStagingRepository"
+
+    try {
+        @Suppress("UNCHECKED_CAST")
+        named(createStagingTaskName) as TaskProvider<CreateStagingRepositoryTask>
+    } catch (e: UnknownTaskException) {
+        register<CreateStagingRepositoryTask>(createStagingTaskName) {
+            group = "publication"
+
+            stagingProfileId.set(ossrhStagingProfileId)
+            user.set(ossrhUsername)
+            password.set(ossrhPassword)
+            repositoryDescription.set("Release v.$version; build ${buildId.get()}")
+            repositoryIdFile.set(repositoryUrlOutputFilePath)
+        }
+    }
+}
+
+tasks.withType<PublishToMavenRepository>().configureEach {
+
+    // https://docs.gradle.org/current/userguide/publishing_customization.html#sec:configuring_publishing_tasks
+    if (name.contains(sonatypeRepoName)) {
+        doFirst {
+
+            // no direct task access, because "cannot be cast to class CreateStagingRepositoryTask" for some reason
+            val repositoryUrl = repositoryUrlOutputFilePath.get().asFile.readText()
```

Comment:
It may be better to do IO in a dedicated task or create a provider.
I don't know a way to implement in `doFirst` / `doLast` a  full contract for a task: caching (inputs/outputs) and so on. 
See `TaskActionWrapper` inside `doFirst` implementation.